keystoremanager/keystoremanager.go

```diff
@@ -155,79 +162,127 @@ attempts to validate the certificate by first checking for an exact match on
 		leafID, err := trustmanager.FingerprintCert(leafCert)
 		if err != nil {
 			logrus.Debugf("error while fingerprinting root certificate with keyID: %s, %v", keyID, err)
 			continue
 		}
 
-		// Check to see if there is an exact match of this certificate.
-		// Checking the CommonName is not required since ID is calculated over
-		// Cert.Raw. It's included to prevent breaking logic with changes of how the
-		// ID gets computed.
-		_, err = km.certificateStore.GetCertificateByKeyID(leafID)
-		if err == nil && leafCert.Subject.CommonName == dnsName {
-			certs[keyID] = rootSigned.Keys[keyID]
+		// Add all the valid leafs to the certificates map so we can refer to them later
+		allCerts[leafID] = leafCert
+
+		// Retrieve all the trusted certificates that match this dns Name
+		certsForCN, err := km.certificateStore.GetCertificatesByCN(dnsName)
+
+		// If there are no certificates with this CN, lets TOFUS!
+		// Note that this logic should only exist in docker 1.8
+		if len(certsForCN) == 0 {
+			km.certificateStore.AddCert(leafCert)
+			certsForCN = append(certsForCN, leafCert)
+			logrus.Debugf("using TOFUS on %s with keyID: %s", dnsName, leafID)
+		}
+
+		// Iterate over all known certificates for this CN and see if any are trusted
+		for _, cert := range certsForCN {
+			// Check to see if there is an exact match of this certificate.
+			certID, err := trustmanager.FingerprintCert(cert)
+			if err == nil && certID == leafID {
+				validKeys[keyID] = rootSigned.Signed.Keys[keyID]
+				logrus.Debugf("found an exact match for %s with keyID: %s", dnsName, keyID)
+			}
 		}
 
-		// Check to see if this leafCertificate has a chain to one of the Root CAs
-		// of our CA Store.
-		certList := []*x509.Certificate{leafCert}
-		err = trustmanager.Verify(km.caStore, dnsName, certList)
+		// Check to see if this leafCertificate has a chain to one of the Root
+		// CAs of our CA Store.
+		err = trustmanager.Verify(km.caStore, dnsName, decodedCerts)
 		if err == nil {
-			certs[keyID] = rootSigned.Keys[keyID]
+			validKeys[keyID] = rootSigned.Signed.Keys[keyID]
+			logrus.Debugf("found a CA path for %s with keyID: %s", dnsName, keyID)
 		}
 	}
 
-	if len(certs) < 1 {
-		return errors.New("could not validate the path to a trusted root")
+	if len(validKeys) < 1 {
+		logrus.Debugf("wasn't able to trust any of the root keys")
+		return ErrValidationFail
+	}
+
+	// TODO(david): change hardcoded minversion on TUF.
+	newRootKey, err := signed.VerifyRoot(root, 0, validKeys, 1)
+	if err != nil {
+		return err
 	}
 
-	_, err = signed.VerifyRoot(root, 0, certs, 1)
+	// VerifyRoot returns a non-nil value if there is a root key rotation happening.
+	// If this happens, we should replace the old root of trust with the new one
+	if newRootKey != nil {
+		logrus.Debugf("got a new root key to rotate to: %s", newRootKey.ID())
+		// Retrieve the certificate associated with the new root key and trust it
+		newRootKeyCert := allCerts[newRootKey.ID()]
```

Comment:
I think we should check to make sure this entry is present in the map. Something could have gone wrong during the loop above where we're supposed to add it to the map.